sdk-extensions/autoconfigure/src/main/java/io/opentelemetry/sdk/autoconfigure/internal/ServiceInstanceIdResourceProvider.java

```diff
@@ -0,0 +1,39 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package io.opentelemetry.sdk.autoconfigure.internal;
+
+import io.opentelemetry.api.common.AttributeKey;
+import io.opentelemetry.api.common.Attributes;
+import io.opentelemetry.sdk.autoconfigure.spi.ConfigProperties;
+import io.opentelemetry.sdk.autoconfigure.spi.ResourceProvider;
+import io.opentelemetry.sdk.autoconfigure.spi.internal.ConditionalResourceProvider;
+import io.opentelemetry.sdk.resources.Resource;
+import java.util.UUID;
+
+/**
+ * does not implement {@link ResourceProvider}, because it depends on all attributes discovered by
+ * the other providers.
+ */
+public final class ServiceInstanceIdResourceProvider implements ConditionalResourceProvider {
```

Comment:
What do you think about changing this to be implemented via `ResourceConfiguration#createEnvironmentResource`? The behavior here is dependent on whether `service.instance.id` is already included via `OTEL_RESOURCE_ATTRIBUTES` and `createEnvironmentResource` already parses that.

The benefit to splitting it out is that this can be independently disabled. But if its split out, then we really want to make sure it runs _after_ `EnvironmentResourceProvider` and so it should return a higher value for `order()`.

Reply:
But I don't have a strong opinion either way.

Reply:
We have two options: 

- Generate `service.instance.id` first (or early)
- Generate `service.instnace.id` last (or late)

If we generate early, resource providers like `EnvironemntResourceProvider` still can override when a user sets `OTEL_RESOURCE_ATTRIBUTES=1234`. But consider a resource provider which runs after a uuid `service.instance.id` is generated but before `EnvironmentResourceProvider` - it will be able to see that `service.instance.id` has been generated, but won't know how high of a priority that id is. In contrast, if we generate a uuid `service.instance.id` last after all other resource providers have run and only if no `service.instance.id` has already bene provided, then any other resource provider which aims to provide a more useful `service.instance.id` can do so after confirming that no previous resource provider has already generated one. 

Example:

**Generate `service.instance.id` first**

1. ServiceInstanceIdResourceProvider provides a uuid `service.instance.id` 
2. A custom FooResourceProvider aims to provide a more durable `service.instance.id`. It sees that `service.instance.id` already exists and isn't sure if the value it provides is higher quality. 
3. EnvironmentResourceProvider runs an overrides `service.instance.id` if its included in `OTEL_RESOURCE_ATTRIBUTES`.

**Generate `service.instance.id` last**

1. A custom FooResourceProvider aims to provide a more durable `service.instance.id`. It doesn't see a `service.instance.id`, so it safely provides a `service.instance.id`.
2. EnvironmentResourceProvider runs an overrides `service.instance.id` if its included in `OTEL_RESOURCE_ATTRIBUTES`.
3. ServiceInstanceIdResourceProvider provides a uuid `service.instance.id` if and only if no `service.instance.id` has been previously provided. 

Generating last seems better - less ambiguity for other resource providers that may want to provide `service.instance.id`.

Reply:
yes, you're right - this is the behavior I wanted - I'll fix it

Reply:
> Generating last seems better - less ambiguity for other resource providers that may want to provide `service.instance.id`.

is this the same way we default `service.name` to `unknown_service:java`?

Reply:
No - the `unknown_service:java` is present when you call `Resource.getDefault()`, which is the [base on top of which](https://github.com/open-telemetry/opentelemetry-java/blob/main/sdk-extensions/autoconfigure/src/main/java/io/opentelemetry/sdk/autoconfigure/ResourceConfiguration.java#L88) all other resource providers are merged. 

The outcome of this decision is interesting:
- A resource provider looking to contribute to `service.name` can't make a decision whether or not to provide it based on the presence of `service.name` alone, since its always present.
- A resource provider CAN reliably check if the present `service.name` is the default by comparing to `Resource.getDefault().getAttribute(SERVICE_NAME)`. 
- This is in contrast to `service.instance.id`. If generated first, then another provider would have no way of identifying that the `service.instance.id` is a randomly generated uuid or some value of higher significance. They can check if it matches the uuid format, but can't know if its random or not.
- Generating `service.instance.id` last does make it hard to remove. Although it is still possible via [AutoConfigurationCustomizer#addResourceCustomizer](https://github.com/open-telemetry/opentelemetry-java/blob/main/sdk-extensions/autoconfigure-spi/src/main/java/io/opentelemetry/sdk/autoconfigure/spi/AutoConfigurationCustomizer.java#L45-L46).